autogen/agentchat/contrib/compressible_groupchat.py

```diff
@@ -0,0 +1,160 @@
+import sys
+from autogen.token_count_utils import count_token, get_max_token_limit, num_tokens_from_functions
+import copy
+
+try:
+    from termcolor import colored
+except ImportError:
+
+    def colored(x, *args, **kwargs):
+        return x
+
+
+from autogen.agentchat.contrib.compressible_agent import CompressibleAgent
+from autogen.agentchat.groupchat import GroupChat
+
+
+class CompressibleGroupChatManager(CompressibleAgent):
+    """(In preview) A chat manager agent that can manage a group chat of multiple agents."""
+
+    def __init__(
+        self,
+        groupchat: GroupChat,
+        name: Optional[str] = "chat_manager",
+        # unlimited consecutive auto reply by default
+        max_consecutive_auto_reply: Optional[int] = sys.maxsize,
+        human_input_mode: Optional[str] = "NEVER",
+        system_message: Optional[str] = "Group chat manager.",
+        # seed: Optional[int] = 4,
+        **kwargs,
+    ):
+        super().__init__(
+            name=name,
+            max_consecutive_auto_reply=max_consecutive_auto_reply,
+            human_input_mode=human_input_mode,
+            system_message=system_message,
+            **kwargs,
+        )
+        self.register_reply(
+            Agent, CompressibleGroupChatManager.run_chat, config=groupchat, reset_config=GroupChat.reset
+        )
+        # self._random = random.Random(seed)
+
+        if self.compress_config:
+            init_count = count_token(
+                groupchat.select_speaker_msg(groupchat.agents), self.llm_config.get("model")
+            ) + count_token(groupchat.selector_end_msg(), self.llm_config.get("model"))
+            trigger_count = self.compress_config["trigger_count"]
+            if init_count >= trigger_count:
+                print(
+                    f"Warning: trigger_count {trigger_count} is less than the initial token count to select speaker {init_count}. Compression will be performed at each turn. Please increase trigger_count if this is not desired."
+                )
+
+    def run_chat(
+        self,
+        messages: Optional[List[Dict]] = None,
+        sender: Optional[Agent] = None,
+        config: Optional[GroupChat] = None,
+    ) -> Union[str, Dict, None]:
+        """Run a group chat."""
+        if messages is None:
+            messages = self._oai_messages[sender]
+        message = messages[-1]
+        speaker = sender
+        groupchat = config
+        for i in range(groupchat.max_round):
+            # set the name to speaker's name if the role is not function
+            if message["role"] != "function":
+                message["name"] = speaker.name
+
+            groupchat.messages.append(message)
+            # broadcast the message to all agents except the speaker
+            for agent in groupchat.agents:
+                if agent != speaker:
+                    self.send(message, agent, request_reply=False, silent=True)
+
+            # the only part changed from vanilla GroupChatManager
```

Comment:
If this is the only part changed, ideally you'll want to remove all the duplication of the rest of the code. Otherwise this file will get obsolete easily when group chat changes, which is already happening with other PRs such as #753. In fact it's already obsolete I think. Are you aware of #688, for example?

Reply:
I am inheriting from `compressible_agent` to build this `CompressibleGroupChatManager`, so it is separated from the original `GroupChatManager`. Since #688 and #753 didn't change `GroupChatManager`, this PR is not obsolete and compatible. 

If `GroupChatManager` got modified, with the current implementation, the only way to do is to modify this compressible version accordingly. 

Another possible way is to inherit from both groupchatmangaer and compressible_agent, but I need to modify the original `GroupChatManager`, do you think this is OK?

Reply:
Composition is better than inheritance. I can't answer your other questions because I don't understand your design. It's worth documenting your design before implementing it.

Reply:
I have change the structure of the the compressiblegroupmanager. I used a hybrid of composition and inheritance, hope this clear things up.

I added a placeholder function "check_groupchat_status" in original groupchat. 
New structure: 
   1. Inherit from GroupChatManager: override the function check_groupchat_status to
        check if the tokens used in groupchat is over the limit, and compress if needed.
   2. Composition with CompressibleAgent: use CompressibleAgent to manage the compression.